### PR TITLE
rust - update to 1.61.0

### PR DESCRIPTION
**Pull request by fritzkink**
Tests are highly appreciated.

### CODE REVIEW

**Comment by AndWac** [comment]
Regarding the use of rustup: Alas it depends on fairly outdated system/library/gcc-4-runtime and system/library/g++-4-runtime which isn't detected because it's a real build dependency. Alas we have somehow broken measures for build, test, and runtime dependencies. We should think about improvements here.

File: components/developer/rust/Makefile [hunk 1]
```diff
@@ -24,44 +24,67 @@
...
+COMPONENT_ARCHIVE_HASH=	sha256:ad0b4351675aa9abdf4c7e066613bd274c4391c5506db152983426376101daed
 COMPONENT_ARCHIVE_URL=	https://static.rust-lang.org/dist/$(COMPONENT_ARCHIVE)
 COMPONENT_LICENSE=	MIT or Apache-2.0
+COMPONENT_LICENSE_FILE=	LICENSE-APACHE
 
-RUST_STAGE0_VER=        1.44.1
-RUST_ARCH:=             x86_64-sun-solaris
+RUST_PRE_VERSION=	1.60.0
+RUST_BOOTSTRAP_PATH=	$(BUILD_DIR)/$(MACH64)
+RUST_ARCH=		x86_64-unknown-illumos
```
Comment (AndWac) [inline components/developer/rust/Makefile:51]: TEST_TARGET= $(NO_TESTS)
should be added as tests seem to be broken.

**Comment by jgmills** [comment]
You probably again have a $HOME/.cargo directory.  The  Makefile line

`COMPONENT_BUILD_ENV =
`
destroys the previous environment settings, including the CARGO_HOME setting just above.  The build falls back to the default because of that line.

**Comment by fritzkink** [comment]
You are right with the $HOME/.cargo directory but without COMPONENT_BUILD_ENV = the build fails immediately, see below

make[1]: Entering directory '/usr/share/src/myoi-userland/components/developer/rust/build/amd64'
make[1]: Nothing to be done for 'x.py'.
Building rustbuild
warning: unused config key `build.cargo` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.docs` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.compiler-docs` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.cargo-native-static` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.configure-args` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.extended` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.vendor` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
warning: unused config key `build.python` in `/usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml`
error: expected a table, but found a string for `target.x86_64-unknown-illumos.llvm-config` in /usr/src/myoi-userland/components/developer/rust/build/amd64/config.toml
failed to run: /usr/src/myoi-userland/components/developer/rust/build/amd64/bin/cargo build --manifest-path /usr/src/myoi-userland/components/developer/rust/rustc-1.61.0-src/src/bootstrap/Cargo.toml --frozen
Build completed unsuccessfully in 0:00:02
make[1]: *** [Makefile:56: install] Error 1
make[1]: Leaving directory '/usr/share/src/myoi-userland/components/developer/rust/build/amd64'
make: *** [/usr/share/src/myoi-userland/make-rules/configure.mk:195: /usr/src/myoi-userland/components/developer/rust/build/amd64/.installed] Error 2

**Comment by jgmills** [comment]
Here's what I've been using.  The display shows all lines in Makefile that contain _ENV .  It's not something that you can apply blindly, but it shows some context:

```
  COMPONENT_PRE_CONFIGURE_ACTION = $(CP) -r $(SOURCE_DIR)/* $(@D)/
  
  # Ignore the previous environment
+ COMPONENT_BUILD_ENV = -i
  
  # Put the bits cargo downloads in a private directory.  This could be cached
  # somewhere more permanent, but it's important to make sure that a person's
  # $HOME/.cargo isn't used.
+ COMPONENT_BUILD_ENV +=	CARGO_HOME=$(BUILD_DIR)/.cargo
  
  # Rust expects the library path to be /usr/lib and is broken otherwise 
  RUSTC_LIBDIR= $(USRLIBDIR)
***************
*** 116,122 ****
--- 118,133 ----
  CONFIGURE_OPTIONS += --enable-llvm-link-shared
  CONFIGURE_OPTIONS += --llvm-root=$(CLANG_ROOT)
  
+ COMPONENT_BUILD_ENV += PATH=$(PATH)
+ COMPONENT_BUILD_ENV += CC=$(CC)
+ COMPONENT_BUILD_ENV += CFLAGS="$(CFLAGS)"
+ COMPONENT_BUILD_ENV += CXX=$(CXX)
+ COMPONENT_BUILD_ENV += CXXFLAGS="$(CXXFLAGS)"
+ COMPONENT_BUILD_ENV += AR=$(GNUAR)
+ COMPONENT_BUILD_ENV += RUSTC=$(RUST_BOOTSTRAP_PATH)/bin/rustc
  # Enforce linker consistency
+ COMPONENT_BUILD_ENV += RUSTFLAGS="-C linker=$(CC)"
+ COMPONENT_BUILD_ENV += RUST_BACKTRACE=1
  
  GMAKE = $(PYTHON)
  COMPONENT_BUILD_GMAKE_ARGS = 
***************
*** 125,131 ****
--- 136,145 ----
  
  COMPONENT_INSTALL_ARGS = ./x.py -v -v -v
  # Cleanup standard environment
+ COMPONENT_INSTALL_ENV = $(COMPONENT_BUILD_ENV)
+ COMPONENT_TEST_ENV = $(COMPONENT_BUILD_ENV)
  # Set install path
+ COMPONENT_INSTALL_ENV += DESTDIR=$(PROTO_DIR)
  
  COMPONENT_PRE_INSTALL_ACTION = mkdir -p $(PROTO_DIR)
  
```

**Pull request merged**